internal/util/basesubjectset.go

```diff
@@ -3,210 +3,230 @@ package util
+	// GetConditionalExpression returns the caveat expression for this subject, if it is conditional.
+	GetConditionalExpression() *v1.CaveatExpression
+
+	// GetExcludedSubjects returns the list of subjects excluded. Must only have values
+	// for wildcards and must never be nested.
+	GetExcludedSubjects() []T
 }
 
-// BaseSubjectSet defines a set that tracks accessible subjects. It is generic to allow
-// other implementations to define the kind of tracking information associated with each subject.
+// BaseSubjectSet defines a set that tracks accessible subjects, their exclusions (if wildcards),
+// and all conditional expressions applied due to caveats.
+//
+// It is generic to allow other implementations to define the kind of tracking information
+// associated with each subject.
 //
 // NOTE: Unlike a traditional set, unions between wildcards and a concrete subject will result
 // in *both* being present in the set, to maintain the proper set semantics around wildcards.
-type BaseSubjectSet[T Subject] struct {
-	values      map[string]T
-	constructor func(subjectID string, excludedSubjectIDs []string, sources ...T) T
-	combiner    func(existing T, added T) T
+type BaseSubjectSet[T Subject[T]] struct {
+	concrete *concreteSubjectSet[T]
+	wildcard *wildcardSubjectTracker[T]
 }
 
 // NewBaseSubjectSet creates a new base subject set for use underneath well-typed implementation.
 //
 // The constructor function is a function that returns a new instancre of type T for a particular
 // subject ID.
-// The combiner function is optional, and if given, is used to combine existing elements in the
-// set into a new element. This is typically used in debug packages for tracking of additional
-// metadata.
-func NewBaseSubjectSet[T Subject](
-	constructor func(subjectID string, excludedSubjectIDs []string, sources ...T) T,
-	combiner func(existing T, added T) T,
-) BaseSubjectSet[T] {
+func NewBaseSubjectSet[T Subject[T]](constructor constructor[T]) BaseSubjectSet[T] {
 	return BaseSubjectSet[T]{
-		values:      map[string]T{},
-		constructor: constructor,
-		combiner:    combiner,
+		concrete: newConcreteSubjectSet[T](constructor),
+		wildcard: newWildcardSubjectTracker[T](constructor),
 	}
 }
 
+// constructor defines a function for constructing a new instance of the Subject type T for
+// a subject ID, its (optional) conditional expression, any excluded subjects, and any sources
+// for bookkeeping. The sources are those other subjects that were combined to create the current
+// subject.
+type constructor[T Subject[T]] func(subjectID string, conditionalExpression *v1.CaveatExpression, excludedSubjects []T, sources ...T) T
+
 // Add adds the found subject to the set. This is equivalent to a Union operation between the
-// existing set of subjects and a set containing the single subject.
+// existing set of subjects and a set containing the single subject, but modifies the set
+// *in place*.
 func (bss BaseSubjectSet[T]) Add(foundSubject T) bool {
```

Comment:
I haven't gotten to the callsites yet, but it's not obvious to me why `Add` should return a bool or why it's helpful to know if the set had a wildcard before the add.

Reply:
standard set semantics: did this subject already exist when I added it